[CIR][Lowering] Core dialects: passthrough symbol visibility

felixdaas opened the pull request:
Fixes #1619 
Fixes #1618

bcardosolopes reviewed (commented):
Awesome, thanks for your first contributions. Can you please a testcase?

bcardosolopes commented on clang/test/CIR/Lowering/ThroughMLIR/compilation.c:
```diff
@@ -0,0 +1,8 @@
+// RUN: %clang -fclangir -fno-clangir-direct-lowering %s -o %t
+
+int main(int argc, char *argv[]) {
```
Can you tide this up a bit? 2 spaces + rename `main` to some other function name.

bcardosolopes replied:
I still see 4 space identation.

bcardosolopes commented on clang/test/CIR/Lowering/ThroughMLIR/declaration.c:
```diff
@@ -0,0 +1,29 @@
+// RUN: split-file %s %t
+// RUN: %clang_cc1 -fclangir -fno-clangir-direct-lowering -emit-mlir=core %t%{fs-sep}test_declaration.c 
+// RUN: FileCheck --input-file=%t%{fs-sep}test_declaration.mlir %t%{fs-sep}test_declaration.c
```
You can merge both files into one, and test all of them in one go, no need to split files.

bcardosolopes commented on clang/test/CIR/Lowering/ThroughMLIR/declaration.c:
```diff
@@ -0,0 +1,29 @@
+
+// CHECK: func.func private @add10(i32) -> i32
+
+int main(int argc, char *argv[]) {
```
Avoid using `main` in tests!

bcardosolopes reviewed (approved):
No problemo, here to help. Thanks for adding the tests - some nits but LGTM once fixed.

felixdaas commented:
Thanks for the feedback! I've implemented the requested changes. Additionally, I added the -triple x86_64-unknown-linux-gnu option to the %clang_cc1 invocation in the declaration.c test, following the pattern used in other ThroughMLIR tests. I also constrained the compilation test to run only on x86_64 Linux targets, so hopefully both will now pass on Windows and macOS as well.

I wasn't entirely sure which other targets to include. Based on [CIRGenModule::getTargetCIRGenInfo()](https://github.com/llvm/clangir/blob/466e3b222ff8f2bce675b17850eb661d20220750/clang/lib/CIR/CodeGen/CIRGenModule.cpp#L509-L558), it might make sense to relax the constraints in the future to support certain aarch64 targets (and possibly spirv64, nvptx, and amdgcn). But for now, I’ve conservatively limited the test to x86_64 Linux only.

bcardosolopes commented on clang/test/CIR/Lowering/ThroughMLIR/compilation.c:
```diff
@@ -0,0 +1,17 @@
+// RUN: %clang -fclangir -fno-clangir-direct-lowering -c %s -o %t
+
+// TODO: Relax target constraint once CIR supports additional backends.
+// REQUIRES: target=x86_64{{.*}}-linux{{.*}}
```
I get what you are trying to do now, please remove this test and just add an extra check to `clang/test/CIR/driver.c` instead. You don't need a `REQUIRES`, invoke clang with the target you are looking for instead (you will see other examples there).

felixdaas replied:
thanks for the info i will do that

bcardosolopes commented on clang/test/CIR/Lowering/ThroughMLIR/declaration.c:
```diff
@@ -0,0 +1,16 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir -fno-clangir-direct-lowering -emit-mlir=core %s -o %t.mlir 
+// RUN: FileCheck --input-file=%t.mlir %s
+
+
```
Rename this test file to `function-attributes.c` and actually add checks for sym_visibility, arg_attrs, etc, which is the goal of this PR.

felixdaas replied:
I have actually thought about it a bit more in regards to arg_attrs and res_attrs (maybe I should have done this earlier). Correct me if I'm wrong: As I understand it they are mostly used in the func dialect to keep information about the attributes and the results operands of the function, which can be used for further analysis, optimizations, etc. Do they have the same purpose on the cir.func and if so does cir use builtin attributes or cir specific attributes to do this? Because if cir specific attributes would be used here it would not be enough to just pass them through but we would also need a mechanism of translating them into builtin attributes. I'm also not sure if there's is a current practical use of passing these attributes to func.func operation. So if you agree I would probably remove the passing through of the arg_attrs and res_attrs attributes for now and it could be implemented in the future if its actually needed.

bcardosolopes replied:
> Do they have the same purpose on the cir.func and if so does cir use builtin attributes or cir specific attributes to do this? Because if cir specific attributes would be used here it would not be enough to just pass them through but we would also need a mechanism of translating them into builtin attributes

It depends on the attribute, some have a direct mapping while others can be inferred, we'll need all of them of some  form in order to emit LLVM that looks like traditional codegen. 

> I'm also not sure if there's is a current practical use of passing these attributes to func.func operation.

Probably not worth bothering for now.

bcardosolopes commented on clang/test/CIR/Lowering/ThroughMLIR/declaration.c:
```diff
@@ -0,0 +1,16 @@
+int declaration_test() {
+    // Variables
+    int number = 15;
+
```
Please use 2 spaces, remove this comment (which doesn't add anything) and remove the newlines:
```
int declaration(int x);
int declaration_test() {
  int number = 15;
  return declaration(number);
}
```

felixdaas commented on clang/test/CIR/Lowering/ThroughMLIR/declaration.c:
```diff
@@ -0,0 +1,16 @@
+
+int declaration(int x);
+
+// CHECK: func.func private @declaration(i32) -> i32
```
the passing of the sym_visibility is actually checked here as the sym_visibility is printed after the func.func (I decided to test it this way since it is pretty similar to the example which actually invokes the error), but I can either extend the test to also check in case of public sym_visibility or replace it with an cir-opt test where cir.func operations are lowered to make it more clear.

bcardosolopes replied:
this is fine, thanks

bcardosolopes commented on clang/test/CIR/driver.c:
```diff
@@ -25,6 +25,7 @@
 // RUN: FileCheck --input-file=%t.cir %s -check-prefix=CIR_MACOS
 // RUN: %clang -target arm64-apple-macosx12.0.0 -fclangir -S -emit-llvm %s -o %t3.ll
 // RUN: FileCheck --input-file=%t3.ll %s -check-prefix=LLVM_MACOS
+// RUN: %clang -target x86_64-unknown-linux-gnu -fclangir -fno-clangir-direct-lowering -c %s -o %t
```
This path is already tested up there in the other -fno-clangir-direct-lowering invocation, turns out you don't need this test.

felixdaas replied:
Maybe I'm missing something here but I don't see where this is already tested. I't looks to me like all the other tests using the indirect lowering "only" create the cir,core dialects or llvm ir and never actually compile to an object file (the error occurs when trying to compile the llvm ir to an object file since during the mlir core to llvm ir translation the cir.triple attribute is not converted to the llvm.triple attribute and is thus discarded). If I try to add this line to the driver.c test and run the test locally (on ubuntu) with the current code on llvm/clangir:main I get the error mentioned in issue #1619 
``` text 
error: unable to create target: 'No available targets are compatible with triple ""'
```
If I run the test locally with the update code of this PR the test passes, so it seems to like like it does add value.

bcardosolopes replied:
True, thanks for clarifying! One last change, use `-o %t.o`

felixdaas commented:
(and btw @bcardosolopes thanks for all the help and info on this PR and sorry about missing your suggestion about the space identation earlier)

bcardosolopes commented on clang/test/CIR/Lowering/ThroughMLIR/function-attributes.c:
```diff
@@ -0,0 +1,9 @@
+int declaration(int x);
+int declaration_test() {
+  return declaration(15);
+}
```
One more nit: code style requires all files to terminate with a newline,
